xdsl/dialects/experimental/dmp.py

```diff
@@ -426,6 +427,160 @@ def print_parameters(self, printer: Printer) -> None:
+    sized segments.
+    """
+
+    name = "dmp.grid_slice_2d"
+
+    topology: ParameterDef[RankTopoAttr]
+
+    diagonals: ParameterDef[builtin.BoolAttr]
+
+    def __init__(self, topo: tuple[int, ...]):
+        super().__init__(
+            [RankTopoAttr(topo), builtin.BoolAttr.from_int_and_width(0, 1)]
+        )
+
+    def _verify(self):
+        assert (
+            len(self.topology.as_tuple()) >= 2
+        ), "GridSlice2d requires at least two dimensions"
+
+    def calc_resize(self, shape: tuple[int, ...]) -> tuple[int, ...]:
+        assert len(shape) >= 2, "GridSlice2d requires at least two dimensions"
+        for size, node_count in zip(shape, self.topology.as_tuple()):
+            assert (
+                size % node_count == 0
+            ), "GridSlice2d requires domain be neatly divisible by shape"
+        return (
+            *(
+                size // node_count
+                for size, node_count in zip(shape, self.topology.as_tuple())
+            ),
+            *(size for size in shape[2:]),
+        )
+
+    def halo_exchange_defs(self, shape: ShapeAttr) -> Iterable[ExchangeDeclarationAttr]:
+        yield from _flat_face_exchanges_for_dim(shape, 0)
+
+        yield from _flat_face_exchanges_for_dim(shape, 1)
+
+        # TOOD: add diagonals
+        assert not self.diagonals
+
+    def comm_layout(self) -> RankTopoAttr:
+        return RankTopoAttr(self.topology.as_tuple())
+
+
+@irdl_attr_definition
+class GridSlice3dAttr(DomainDecompositionStrategy):
+    """
+    Takes a grid with two or more dimensions, slices it along the first three.
+    """
+
+    name = "dmp.grid_slice_3d"
+
+    topology: ParameterDef[RankTopoAttr]
+
+    diagonals: ParameterDef[builtin.BoolAttr]
+
+    def __init__(self, topo: tuple[int, ...]):
+        super().__init__(
+            [RankTopoAttr(topo), builtin.BoolAttr.from_int_and_width(0, 1)]
+        )
+
+    def _verify(self):
+        assert (
+            len(self.topology.as_tuple()) >= 3
+        ), "GridSlice3d requires at least three dimensions"
+
+    def calc_resize(self, shape: tuple[int, ...]) -> tuple[int, ...]:
+        assert len(shape) >= 3, "GridSlice3d requires at least two dimensions"
+        for size, node_count in zip(shape, self.topology.as_tuple()):
+            assert (
+                size % node_count == 0
+            ), "GridSlice3d requires domain be neatly divisible by shape"
+        return (
+            *(
+                size // node_count
+                for size, node_count in zip(shape, self.topology.as_tuple())
+            ),
+            *(size for size in shape[3:]),
+        )
+
+    def halo_exchange_defs(self, shape: ShapeAttr) -> Iterable[ExchangeDeclarationAttr]:
+        yield from _flat_face_exchanges_for_dim(shape, 0)
+
+        yield from _flat_face_exchanges_for_dim(shape, 1)
+
+        yield from _flat_face_exchanges_for_dim(shape, 2)
+
+        # TOOD: add diagonals
```

Comment:
TOOD!

Reply:
I thought you ondly didn't like TODOs? :stuck_out_tongue_winking_eye: 

Just moved the code, will polish that